### PR TITLE
Loading images

### PR DESCRIPTION
**Pull request by wagenrace**
Issue #16

The following pull request adds, fixes, and/or modifies:

Add loading of image and relate it to data of the csv file

I performed the following prior to filing this pull request:

- [X] Tested that my change does not break the analysis pipeline
- [X] Ran a linter through my code
- [X] Update environment dependencies if my code introduces a new package

### CODE REVIEW

File: utils/load_image.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    replicate: int = 1,
+    wellName: str = "C10",
+    field: int = 1,
+) -> Image:
+    """
+    Find the single image that matches on all given parameters.
+
+    Returns: image
+
+    WARNING:
```
Comment (gwaygenomics) [inline utils/load_image.py]: this is amazing!

File: utils/load_image.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    )
+    b = extracted_file.read()
+
+    img = Image.open(BytesIO(b))
+    return img
+
+
+def get_all_images(dataframe: pd.dataframe) -> (Image, str, str):
+    """
+    Creates a generator with all the image that match exactly one row in dataframe.
```
Comment (gwaygenomics) [inline utils/load_image.py]: What dataframe are you talking about?

Comment (wagenrace) [reply utils/load_image.py]: The variable called dataframe, what needs to be a panda dataframe

**Comment by wagenrace** [comment]
@gwaygenomics and @jccaicedo could any of you run it on your computer? I would love to merge it but want to prevent the 'it work on my computer' bug

File: environment.yml [hunk 3]
```diff
@@ -11,6 +11,7 @@ dependencies:
 - conda-forge::scikit-learn=0.21.3
 - conda-forge::black=19.10b0
 - conda-forge::seaborn=0.9.0
+- conda-forge::pillow=4.1.2
```
Comment (gwaygenomics) [inline environment.yml]: I got a conda install error for this version of pillow.

```sh
Solving environment: failed

ResolvePackageNotFound: 
  - conda-forge::pillow=4.1.2
```

Adding [pillow=6.2.1](https://anaconda.org/conda-forge/pillow/files) seems to have worked.

Will this break the pipeline?

File: utils/load_image.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    field: int = 1,
+) -> Image:
+    """
+    Find the single image that matches on all given parameters.
+
+    Returns: image
+
+    WARNING:
+    Use this only if you need to load a single image.
+    Every time you run this function it loops over the whole zip file.
```
Comment (gwaygenomics) [inline utils/load_image.py]: This is great docstring, but I am wondering if there any reason why it needs to loop over the whole zip file?

Should there be a processing step where all single cell images are extracted first?

File: utils/load_image.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    If you need all images use get_all_images.
+    """
+    newPlateName = plate.replace("P", "S")
+    extracted_file = TARFILE.extractfile(
+        f"training/{target}/211_11_17_X_Man_LOPAC_X5_LP_{newPlateName}_{replicate}_{wellName}_{field}_{cell_id}.tiff"
```
Comment (gwaygenomics) [inline utils/load_image.py]: I am tempted to suggest that we should add some sort of error message and flow control to these arguments.

The error message would catch incorrect inputs and throw an error message of what input arguments are possible.

File: utils/load_image.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import pandas as pd
+
+
+basefolder_loc = Path(__file__).parents[1]
+TARFILE = tarfile.open(
+    os.path.join(basefolder_loc, "1.download-data", "data", "training.tar.gz"), "r:gz",
+)
+
+
+def load_img(
```
Comment (gwaygenomics) [inline utils/load_image.py]: I have tested this function with:

```python
from utils.load_image import load_img
load_img()
```

output:

image

Comment (gwaygenomics) [reply utils/load_image.py]: I also tested:

```python
load_img(
    target="dopaminereceptor",
    plate="P1",
    cell_id=5,
    replicate=2,
    wellName="B11",
    field=1
)
```

output: 

image

File: utils/load_image.py (hunk 5 above)
Comment (gwaygenomics) [reply utils/load_image.py]: In testing the function, I am more convinced that some sort of error control is warranted. Although, do you think users (mainly, us) will interact with this function?

File: utils/load_image.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    )
+    b = extracted_file.read()
+
+    img = Image.open(BytesIO(b))
+    return img
+
+
+def get_all_images(dataframe: pd.DataFrame) -> (Image, str, str):
+    """
+    Creates a generator with all the image that match exactly one row in dataframe.
```
Comment (gwaygenomics) [inline utils/load_image.py]:
```suggestion
    Creates a generator with all the images that match exactly one row in the dataframe.
```

still not sure what the "dataframe" is supposed to be. Maybe a more descriptive variable would help?

Also, how does one use this function? Maybe some usage notes would be helpful

File: utils/load_image.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    The check is done based on target, cell_id, well, plate, field and replicate
+
+    returns: img, cell_code, target
+    """
+    for member in TARFILE.getmembers():
```
Comment (gwaygenomics) [inline utils/load_image.py]: would this logic break if we decide that extracting images from compressed file first is preferred?

File: utils/load_image.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            & (dataframe["field"] == int(field))
+            & (dataframe["replicate"] == int(replicate))
+        ]
+        if len(rows) == 0:
+            continue
+        elif len(rows) > 1:
+            print("To many rows", rows)
+
+        extracted_file = TARFILE.extractfile(member)
+        b = extracted_file.read()
```
Comment (gwaygenomics) [inline utils/load_image.py]: is there any way we could instead call the previous function `load_img()` here instead of duplicating code?

**Comment by gwaygenomics** [comment]
One additional note:

I needed to merge in this branch to master as i was getting strange import errors. 479b578a3f229921e596f1d92523979ae5230647 fixed this.

File: environment.yml (hunk 3 above)
Comment (gwaygenomics) [reply environment.yml]: Added in de28f65450f49b917de78451b4eda8226e944edb. Still worth noting if the upgraded version breaks anything :)

File: utils/load_image.py (hunk 6 above)
Comment (wagenrace) [reply utils/load_image.py]: That looks good! (although they are so dark I needed a moment to see there are not completely black)

File: utils/load_image.py (hunk 8 above)
Comment (wagenrace) [reply utils/load_image.py]: Yes

File: utils/load_image.py (hunk 9 above)
Comment (wagenrace) [reply utils/load_image.py]: Tried that, but everything is extremely slow. The looking up of a file in a compressed tarfile seems to be very slow.
So the two function deal with it differently:
- load_img -> looks up image from tarfile
- get_all_images -> loads all images with looking up any of them and check if there in the list afterward

File: environment.yml (hunk 3 above)
Comment (wagenrace) [reply environment.yml]: Thanks completely overlooked it

File: utils/load_image.py (hunk 4 above)
Comment (wagenrace) [reply utils/load_image.py]: We can first unzip the files but in that case the user needs more memory and load more from its cold to hot memory. I haven't tested if it is really slower but I think so

**Pull request merged**